platform/admin/core/admin-core-impl/src/main/java/org/codice/ddf/admin/core/impl/ConfigurationAdminImpl.java

```diff
@@ -632,64 +613,109 @@ public Map getFactoryPidObjectClasses() {
                 }
             }
         }
         return metatypeList;
     }
 
-    private List<Map<String, Object>> createMetatypeMap(AttributeDefinition[] definitions) {
-        List<Map<String, Object>> metatypeList = new ArrayList<Map<String, Object>>();
+    private List<MetatypeAttribute> createMetatypeMap(AttributeDefinition[] definitions) {
+        List<MetatypeAttribute> metatypeList;
 
         if (definitions != null) {
-            for (AttributeDefinition definition : definitions) {
-                Map<String, Object> attributeMap = new HashMap<String, Object>();
-                attributeMap.put(MAP_ENTRY_ID, definition.getID());
-                attributeMap.put(MAP_ENTRY_NAME, definition.getName());
-                attributeMap.put(MAP_ENTRY_CARDINALITY, definition.getCardinality());
-                attributeMap.put(MAP_ENTRY_DEFAULT_VALUE, definition.getDefaultValue());
-                attributeMap.put(MAP_ENTRY_DESCRIPTION, definition.getDescription());
-                attributeMap.put(MAP_ENTRY_TYPE, definition.getType());
-                attributeMap.put(MAP_ENTRY_OPTION_LABELS, definition.getOptionLabels());
-                attributeMap.put(MAP_ENTRY_OPTION_VALUES, definition.getOptionValues());
-                metatypeList.add(attributeMap);
-            }
+            metatypeList = Arrays.stream(definitions)
+                    .map(MetatypeAttributeImpl::new)
+                    .collect(Collectors.toList());
+        } else {
+            metatypeList = new ArrayList<>();
         }
 
         return metatypeList;
     }
 
-    public boolean isPermittedToViewService(String servicePid) {
+    public boolean isPermittedToViewService(String servicePid, Subject subject) {
         KeyValueCollectionPermission serviceToCheck = new KeyValueCollectionPermission(
                 "view-service.pid",
                 new KeyValuePermission("service.pid", Sets.newHashSet(servicePid)));
-        return SecurityUtils.getSubject()
-                .isPermitted(serviceToCheck);
+        return subject.isPermitted(serviceToCheck);
+    }
+
+    public boolean isPermittedToViewService(String servicePid) {
+        return isPermittedToViewService(servicePid, SecurityUtils.getSubject());
+    }
+
+    public ConfigurationStatus disableManagedServiceFactoryConfiguration(String servicePid,
+            Configuration originalConfig) throws IOException {
+        Dictionary<String, Object> properties = originalConfig.getProperties();
+        String originalFactoryPid =
+                (String) properties.get(org.osgi.service.cm.ConfigurationAdmin.SERVICE_FACTORYPID);
+        if (originalFactoryPid == null) {
+            throw new IOException("Configuration does not belong to a manged service factory.");
+        }
+        if (StringUtils.endsWith(originalFactoryPid, ConfigurationStatus.DISABLED_EXTENSION)) {
+            throw new IOException("Configuration is already disabled.");
```

Comment:
❓ Should we throw `IllegalArgumentException` or `IllegalStateException` here and above? `IOException` doesn't feel like the right one to use.